File: vm/picovm_run.c
/* picovm_run.c -- host test harness for picovm.c.
 *
 * Reads a program from stdin:  first line = word count N, then N lines each a
 * hex 32-bit instruction word.  Runs it and prints final state as lines:
 *
 *   STEPS <n>
 *   FAULT <code> <pc> <detail>
 *   STATUS <http_status>
 *   REGS <r0> <r1> ... <r15>
 *   OUT  <hex bytes>
 *
 * Used by tools to compare against picoscript_vm.PicoVM.
 */
#include "picovm.h"
#include <stdio.h>
#include <stdlib.h>

int main(void)
{
    int n = 0;
    if (scanf("%d", &n) != 1) return 1;

    static uint32_t prog[65536];
    if (n < 0) n = 0;
    if (n > 65536) n = 65536;
    for (int i = 0; i < n; i++) {
        unsigned w = 0;
        if (scanf("%x", &w) != 1) return 1;
        prog[i] = (uint32_t)w;
    }

    static uint8_t arena[520 * 1024];   /* data arena for Memory.* / Io (Pico 2 SRAM size) */
    pv_ctx ctx;
    pv_init(&ctx);
    ctx.mem = arena;
    ctx.mem_size = (long)sizeof(arena);
    const char *ms = getenv("PICOVM_MAX_STEPS");   /* let tests drive the step budget */
    if (ms && *ms) ctx.max_steps = atol(ms);
    const char *cp = getenv("PICOVM_CAPS");         /* let tests restrict binding capabilities */
    if (cp && *cp) ctx.caps = (uint32_t)strtoul(cp, 0, 0);
    const char *na = getenv("PICOVM_NOALLOC");       /* let tests enable hot-path no-alloc mode */
    if (na && *na && na[0] != '0') ctx.no_alloc = 1;
    const char *sd = getenv("PICOVM_SEED");         /* let tests pin Random.U32 nondeterminism */
    if (sd && *sd) ctx.rng_state = (uint64_t)strtoull(sd, 0, 0);
    long steps = pv_vm_run(&ctx, prog, n);

    printf("STEPS %ld\n", steps);
    printf("FAULT %d %d %d\n", ctx.fault, ctx.fault_pc, ctx.fault_detail);
    printf("STATUS %d\n", ctx.http_status);
    printf("REGS");
    for (int i = 0; i < PV_NUM_REGS; i++) printf(" %d", ctx.regs[i]);
    printf("\n");
    printf("OUT");
    for (int i = 0; i < ctx.out_len; i++) printf(" %02x", ctx.out[i]);
    printf("\n");
    return 0;
}

File: vm/picovm_run.js
// picovm_run.js -- Node harness for picovm.js (parity testing vs Python/C VM).
// Reads: first line word count N, then N hex words. Prints STEPS/FAULT/STATUS/REGS/OUT.
const PicoVM = require("./picovm.js");

let data = "";
process.stdin.on("data", (d) => (data += d));
process.stdin.on("end", () => {
  const toks = data.split(/\s+/).filter((s) => s.length);
  const n = parseInt(toks[0], 10);
  const words = [];
  for (let i = 0; i < n; i++) words.push(parseInt(toks[1 + i], 16) >>> 0);
  const ms = process.env.PICOVM_MAX_STEPS;   // let tests drive the step budget
  const cps = process.env.PICOVM_CAPS;        // let tests restrict binding capabilities
  const seed = process.env.PICOVM_SEED;        // let tests pin Random.U32 nondeterminism
  const noalloc = process.env.PICOVM_NOALLOC;   // let tests enable hot-path no-alloc mode
  var opts = {};
  if (ms) opts.maxSteps = parseInt(ms, 10);
  if (cps) opts.caps = parseInt(cps, 0) >>> 0;
  if (seed) opts.seed = parseInt(seed, 0) >>> 0;
  if (noalloc && noalloc[0] !== "0") opts.noAlloc = true;
  const vm = new PicoVM(opts);
  let fault = 0;
  let faultPc = 0;
  let faultDetail = 0;
  try {
    vm.run(words);
  } catch (e) {
    // Map the VM trap to a typed fault code, mirroring the C runtime's ctx.fault.
    faultPc = (e && e.pc) || 0;
    faultDetail = (e && e.detail) || 0;
    var m = String(e && e.message || e);
    if (e && e.fault !== undefined) fault = e.fault;
    else if (m.indexOf("step budget") >= 0) fault = 1;
    else if (m.indexOf("bad opcode") >= 0) fault = 2;
    else if (m.indexOf("bad jump") >= 0 || m.indexOf("bad branch") >= 0 || m.indexOf("bad call") >= 0) fault = 3;
    else if (m.indexOf("template depth") >= 0) fault = 7;
    else if (m.indexOf("capability denied") >= 0) fault = 8;
    else fault = 99;
  }
  console.log("STEPS " + vm.steps);
  console.log("FAULT " + fault + " " + faultPc + " " + faultDetail);
  console.log("STATUS " + vm.httpStatus);
  console.log("ASSERT " + ((vm._asTotal || 0) >>> 0) + " " + ((vm._asFailed || 0) >>> 0));
  console.log("REGS " + Array.from(vm.regs).join(" "));
  console.log("OUT " + vm.output.map((b) => b.toString(16).padStart(2, "0")).join(" "));
});
